test/hotspot/jtreg/gc/TestJNICriticalStressTest.java

```diff
@@ -0,0 +1,236 @@
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                    return;
+                }
+
+                long gcCountBefore = collector.getCollectionCount();
+                wb.fullGC();
+                long gcCountAfter = collector.getCollectionCount();
+
+                Asserts.assertLessThan(gcCountBefore, gcCountAfter, "Triggered more Full GCs than expected");
+            }
+        }
+    }
+
+    static public Map<Integer,String> largeMap;
+
+    public static void main(String... args) throws Exception {
+        if (args.length < 4) {
+            println("usage: JNICriticalStressTest <duration sec> <alloc threads> <jni critical threads> <gc name>");
+            exit(-1);
+        }
+
+        long durationSec = Long.parseLong(args[0]);
+        int allocThreadNum = Integer.parseInt(args[1]);
+        int jniCriticalThreadNum = Integer.parseInt(args[2]);
+
+        StringBuilder OldGCName = new StringBuilder();
+        switch (args[3]) {
+            case "G1":
+                OldGCName.append("G1 Old Generation");
+                break;
+            case "Parallel":
+                OldGCName.append("PS MarkSweep");
+                break;
+            case "Serial":
+                OldGCName.append("MarkSweepCompact");
+                break;
+            default:
+                throw new RuntimeException("Unsupported GC selected");
+        }
+
+        List<GarbageCollectorMXBean> collectors = ManagementFactory.getGarbageCollectorMXBeans();
+
+        for (int i = 0; i < collectors.size(); i++) {
+            if (collectors.get(i).getName().contains(OldGCName.toString())) {
+                collector = collectors.get(i);
+                break;
+            }
+        }
+
+        if (collector == null) {
+            throw new RuntimeException(OldGCName.toString() + " not found");
+        }
+
+        println("Running for " + durationSec + " secs");
+
+        largeMap = populateMap(LARGE_MAP_SIZE);
+
+        println("Starting " + allocThreadNum + " allocating threads");
+        for (int i = 0; i < allocThreadNum; i += 1) {
+            new Thread(new AllocatingWorker()).start();
+        }
+
+        println("Starting " + jniCriticalThreadNum + " jni critical threads");
+        for (int i = 0; i < jniCriticalThreadNum; i += 1) {
+            new Thread(new JNICriticalWorker()).start();
+        }
+
+        new Thread(new SystemGCWorker()).start();
+
+        long durationMS = (long) (1000 * durationSec);
+        long start = System.currentTimeMillis();
+        long now = start;
+        long soFar = now - start;
+        while (soFar < durationMS) {
+            try {
+                Thread.sleep(durationMS - soFar);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+                return;
+            }
+            now = System.currentTimeMillis();
+            soFar = now - start;
+        }
```

Comment:
A single `Thread.sleep` should be fine -- the error margin should be negligible.